lib/web_ui/lib/src/engine/compositor/skia_object_cache.dart

```diff
@@ -0,0 +1,238 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+/// A cache of Skia objects whose memory Flutter manages.
+///
+/// When using Skia, Flutter creates Skia objects which are allocated in
+/// WASM memory and which must be explicitly deleted. In the case of Flutter
+/// mobile, the Skia objects are wrapped by a C++ class which is destroyed
+/// when the associated Dart object is garbage collected.
+///
+/// On the web, we cannot tell when a Dart object is garbage collected, so
+/// we must use other strategies to know when to delete a Skia object. Some
+/// objects, like [ui.Paint], can safely delete their associated Skia object
+/// because they can always recreate the Skia object from data stored in the
+/// Dart object. Other objects, like [ui.Picture], can be serialized to a
+/// JS-managed data structure when they are deleted so that when the associated
+/// object is garbage collected, so is the serialized data.
+class SkiaObjectCache {
+  final int maximumSize;
+
+  final DoubleLinkedQueue<SkiaObject> _itemQueue;
+  final Map<SkiaObject, DoubleLinkedQueueEntry<SkiaObject>> _itemMap;
+
+  SkiaObjectCache(this.maximumSize)
+      : _itemQueue = DoubleLinkedQueue<SkiaObject>(),
+        _itemMap = <SkiaObject, DoubleLinkedQueueEntry<SkiaObject>>{};
+
+  int get length => _itemQueue.length;
+
+  bool contains(SkiaObject object) {
```

Comment:
This linearly scans the queue. Is there a way to make it faster? Although, it seems to be only used in tests. Maybe call it `debugContains` so we don't try to use it in production? `@visibleForTesting` may also be useful here.

Reply:
Done.